text/0037-anr-rates.md

```diff
@@ -0,0 +1,107 @@
+    (uniqCombined64MergeIf((value AS _snuba_value), equals((metric_id AS _snuba_metric_id), 'user')) AS `_snuba_count_unique(sentry.sessions.user)`), 
+    (divide(`_snuba_session.anr_user`,`_snuba_count_unique(sentry.sessions.user)`) AS `_snuba.anr_rate`)
+   FROM metrics_sets_v2_local
+  WHERE equals(granularity, ...)
+    AND equals((org_id AS _snuba_org_id), ...)
+    AND in((project_id AS _snuba_project_id), [...])
+    AND greaterOrEquals((timestamp AS _snuba_timestamp), toDateTime('...', 'Universal'))
+    AND less(_snuba_timestamp, toDateTime('...', 'Universal'))
+    AND in(_snuba_metric_id, [user])
+  LIMIT 2
+```
+
+1. This is a more generic solution that’s extensible to "freeze mechanisms" (ex. app hangs on iOS) from other operating systems
+2. This can also be extended to track the four different types of ANRs outlined [here](https://developer.android.com/topic/performance/vitals/anr). Differentiating ANRs due to `Input dispatching timed out` separately will allow us to calculate user-perceived ANR rate
+3. Stand-alone addition that won’t require any migration, less risky, doesn’t affect existing calculations
+
+## Drawbacks
+
+1. Increases the cardinality of the sessions data, however this tag would only be set by Android SDKs for now, so would only affect Android sessions
+
+## Introduce a new session.status tag value (Option 2)
+
+SDK sends a session update with (`session.status:frozen`) when it hits an ANR in addition to creating an ANR event and we calculate ANR rate with a query similar to counting abnormal or errored users:
+
+```sql
+SELECT 
+    (uniqCombined64MergeIf((value AS _snuba_value), equals((arrayElement(tags.value, indexOf(tags.key, 'session.status')) AS `_snuba_tags[session.status]`), 'frozen') AND in((metric_id AS _snuba_metric_id), [user])) AS `_snuba_session.anr_user`), 
+    (uniqCombined64MergeIf((value AS _snuba_value), equals((metric_id AS _snuba_metric_id), 'user')) AS `_snuba_count_unique(sentry.sessions.user)`), 
+    (divide(`_snuba_session.anr_user`,`_snuba_count_unique(sentry.sessions.user)`) AS `_snuba.anr_rate`)
+   FROM metrics_sets_v2_local
+  WHERE equals(granularity, ...)
+    AND equals((org_id AS _snuba_org_id), ...)
+    AND in((project_id AS _snuba_project_id), [...])
+    AND greaterOrEquals((timestamp AS _snuba_timestamp), toDateTime('...', 'Universal'))
+    AND less(_snuba_timestamp, toDateTime('...', 'Universal'))
+    AND in(_snuba_metric_id, [user])
+  LIMIT 2
+```
+
+## Drawbacks
+
+1. We are introducing a new terminal session state but the android system displays a dialogue asking the user if they would like to wait or force quit the app. So an ANR doesn't necessarily mean a terminating session state. Downstream session status changes will no longer be captured.
```

Comment:
1. there is also already a status for the session update we send when an error happens within the application: `errored`. Are we saying that ANR errors do not mark a session as errored anymore? If ANR does not become its own terminal state, do we still consider the session crashed if it contained an ANR at some point?

2. if ANR is not terminal, how do you deal with multiple ANR in the same session? (that is a question for both proposals)

   i assume it doesn't matter for user count at a high level, but we need to figure out if we want to send multiple session updates

Reply:
ANRs don't change the `status` of the session, but rather increase the `errorCount`, the BE or FE then displays as `Errored`, since errored actually isn't a valid status.

Reply:
Btw if the Android team works on https://github.com/getsentry/sentry-java/issues/1796 which is planned for this Q, this could affect the ANR rate calculation, let's assume that we go for the approach that only reports ANRs on App restart, that means that the session status was actually `crashed` and not only with an `errorCount`.